src/Agent.Listener/Environment.cs

```diff
@@ -0,0 +1,338 @@
+using Microsoft.VisualStudio.Services.Agent.Util;
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.VisualStudio.Services.Agent.Listener
+{
+    [ServiceLocator(Default = typeof(Environment))]
+    public interface IEnvironment : IAgentService
+    {
+        Task<Dictionary<string, string>> GetCapabilities(string agentName, CancellationToken token);
+
+        void EnsureEnvFile();
+
+        IDictionary<string, string> GetEnv();
+    }
+
+    struct Capability
+    {
+        public Capability(string name, string tool, string[] paths)
+        {
+            Name = name;
+            Tool = tool;
+            Paths = paths;
+        }
+
+        //Name of the capability as requried by the task definition.
+        public string Name { get; private set; }
+
+        //Name of an executable file, which if found in the file system means the capability is present.
+        public string Tool { get; private set; }
+
+        //An array of paths, which if found in the file system means the capability is present.
+        public string[] Paths { get; private set; }
+    }
+
+    struct ShellCapability
+    {
+        public ShellCapability(string name, string command, string commandArgs)
+        {
+            Name = name;
+            Command = command;
+            CommandArgs = commandArgs;
+        }
+
+        //Name of the capability as requried by the task definition.
+        public string Name { get; private set; }
+
+        //Name of an executable file, which can determine if a capability is present.
+        //The script is expected to print on the standard output the value for the capability.
+        public string Command { get; private set; }
+
+        //Arguments passed to Command.
+        public string CommandArgs { get; private set; }
+    }
+
+    //extensions which simplify adding constant capabilities into a list
+    static class ListExtensions
+    {
+        public static void Add(this IList<Capability> list, string name, string tool = null, string[] paths = null)
+        {
+            list.Add(new Capability(name, tool, paths));
+        }
+
+        public static void Add(this IList<ShellCapability> list, string name, string command, string commandArgs = null)
+        {
+            list.Add(new ShellCapability(name, command, commandArgs));
+        }
+    }
+
+    public sealed class Environment : AgentService, IEnvironment
+    {
+        private List<Capability> _regularCapabilities =
+            new List<Capability>
+            {
+                { "ant" },
```

Comment:
I'm removing extension methods and using "new Capability", "new ShellCapability" instead.